src/epics.c

```diff
@@ -0,0 +1,551 @@
+  }
+}
+
+static void handle_conn_up(struct pv *p) {
+  int ret = 0;
+
+  if (p->eid) {
+    INFO("epics plugin: Channel \"%s\" reconnected", p->name);
+
+    p->is_active = 1;
+
+    return;
+  }
+
+  if (p->is_label) {
+    p->ch_type = DBR_STRING;
+  } else {
+    const data_set_t *ds = plugin_get_ds(p->value.type);
+
+    if (ds == NULL) {
+      ERROR("epics plugin: Unknown type \"%s\" for channel \"%s\". See "
+            "types.db(5) for details.",
+            p->value.type, p->name);
+
+      return;
+    }
+
+    const int ds_type = ds->ds[0].type;
+
+    p->ch_type = ca_field_type(p->id);
+    if (ds_type == DS_TYPE_COUNTER &&
+        (p->ch_type == DBR_SHORT || p->ch_type == DBR_LONG)) {
+      p->ch_type = DBR_LONG;
+    } else if (ds_type == DS_TYPE_GAUGE &&
+               (p->ch_type == DBR_FLOAT || p->ch_type == DBR_DOUBLE ||
+                p->ch_type == DBR_ENUM)) {
+      p->ch_type = DBR_DOUBLE;
+    } else if (ds_type == DS_TYPE_DERIVE &&
+               (p->ch_type == DBR_SHORT || p->ch_type == DBR_LONG)) {
+      p->ch_type = DBR_LONG;
+    } else if (ds_type == DS_TYPE_ABSOLUTE &&
+               (p->ch_type == DBR_SHORT || p->ch_type == DBR_LONG)) {
+      p->ch_type = DBR_LONG;
+    } else {
+      ERROR("epics plugin: Variable type \"%s\" doesn't match channel type "
+            "\"%s\" for channel \"%s\"",
+            DS_TYPE_TO_STRING(ds_type), dbf_type_to_text(p->ch_type), p->name);
+
+      return;
+    }
+
+    if (ca_element_count(p->id) != ds->ds_num) {
+      ERROR("epics plugin: Variable element number %lu doesn't channel element "
+            "count %lu for channel \"%s\"",
+            ds->ds_num, ca_element_count(p->id), p->name);
+
+      return;
+    }
+
+    p->value.values_len = ds->ds_num;
+    p->value.values = calloc(p->value.values_len, sizeof(value_t));
+    if (p->value.values == NULL) {
+      ERROR("epics plugin: Cannot allocate memory for %lu values or channel "
+            "\"%s\"",
+            ds->ds_num, p->name);
+
+      return;
+    }
+  }
+
+  ret =
+      ca_create_subscription(p->ch_type, ca_element_count(p->id), p->id,
+                             DBE_VALUE | DBE_ALARM, event_handler, p, &p->eid);
+  if (ret != ECA_NORMAL) {
+    ERROR("epics plugin: CA error %s occurred while trying to create "
+          "subscription for channel \"%s\"",
+          ca_message(ret), p->name);
+
+    return;
+  }
+
+  p->is_active = 1;
+}
+
+static void handle_conn_down(struct pv *p) {
+  WARNING("epics plugin: Channel \"%s\" disconnected", p->name);
+
+  p->is_active = 0;
+}
+
+static void connection_handler(struct connection_handler_args args) {
+  struct pv *p = (struct pv *)ca_puser(args.chid);
+
+  switch (args.op) {
+  case CA_OP_CONN_UP:
+    handle_conn_up(p);
+    break;
+  case CA_OP_CONN_DOWN:
+    handle_conn_down(p);
+    break;
+  }
+}
+
+static void *epics_thread(void *args) {
+  int i = 0;
+  long ret = 0;
+
+  ret = ca_context_create(ca_disable_preemptive_callback);
```

Comment:
Please declare variables as late as possible, ideally initializing them to their final value. Here and throughout.

```suggestion
  long ret = ca_context_create(ca_disable_preemptive_callback);
```